staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -386,7 +391,62 @@ func (s *store) GetToList(ctx context.Context, key string, resourceVersion strin
 		return err
 	}
 	// update version with cluster level revision
-	return s.versioner.UpdateList(listObj, uint64(getResp.Header.Revision))
+	return s.versioner.UpdateList(listObj, uint64(getResp.Header.Revision), "")
+}
+
+// continueToken is a simple structured object for encoding the state of a continue token.
+// TODO: if we change the version of the encoded from, we can't start encoding the new version
+// until all other servers are upgraded (i.e. we need to support rolling schema)
+// TODO: be very careful about changing this
+type continueToken struct {
+	APIVersion      string `json:"v"`
+	ResourceVersion int64  `json:"rv"`
+	StartKey        string `json:"start"`
+}
+
+// parseFrom transforms an encoded predicate from into a versioned struct.
+// TODO: return a typed error that instructs clients that they must relist
+func decodeContinue(continueValue, keyPrefix string) (fromKey string, rv int64, err error) {
+	data, err := base64.RawURLEncoding.DecodeString(continueValue)
+	if err != nil {
+		return "", 0, fmt.Errorf("continue key is not valid: %v", err)
+	}
+	var c continueToken
+	if err := json.Unmarshal(data, &c); err != nil {
+		return "", 0, fmt.Errorf("continue key is not valid: %v", err)
+	}
+	switch c.APIVersion {
+	case "v1":
+		if c.ResourceVersion == 0 {
+			return "", 0, fmt.Errorf("continue key is not valid: incorrect encoded start resourceVersion (version v1)")
+		}
+		if len(c.StartKey) == 0 {
+			return "", 0, fmt.Errorf("continue key is not valid: encoded start key empty (version v1)")
+		}
+		cleaned := path.Clean(c.StartKey)
```

Comment:
I'm wondering about this one. Are the paths that are "Clean-equivalent" really equivalent in etcd?

Reply:
For the purposes of this code, it's mostly ensuring no path traversal attacks are possible.  If the start key is `foo/bar/baz` and this is a cluster scoped resource, it doesn't matter.  If an attacker passes `/`, we still append to the end of the current prefix.

Reply:
Got it - can you please add a comment about it. Your goal definitely wasn't clear to me before you explained it.

Reply:
This cannot happen as etcd doesn't have idea of path. Key is just a blob, it's Kubernetes that puts paths in key.

Reply:
etcd may not have paths, but by using slashes and prefixing, we have paths and traversal attacks #113756

Reply:
This applies to apiserver HTTP server that will interpret `/safePath/../unsafePath` as `/unsafePath`. However etcd doesn't have concept of path and will interpret `/safePath/../unsafePath` as `/safePath/../unsafePath`.

So no, etcd is not vulnerable to traversal attacks.